ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/SqlLedger.scala

```diff
@@ -334,33 +319,70 @@ private class SqlLedgerFactory(ledgerDao: LedgerDao) {
   private def reset(): Future[Unit] =
     ledgerDao.reset()
 
-  private def initialize(initialLedgerId: Option[String]): Future[String] = initialLedgerId match {
-    case Some(initialId) =>
-      ledgerDao
-        .lookupLedgerId()
-        .flatMap {
-          case Some(foundLedgerId) if (foundLedgerId == initialId) =>
-            ledgerFound(foundLedgerId)
-          case Some(foundLedgerId) =>
-            val errorMsg =
-              s"Ledger id mismatch. Ledger id given ('$initialId') is not equal to the existing one ('$foundLedgerId')!"
-            logger.error(errorMsg)
-            sys.error(errorMsg)
-          case None =>
-            doInit(initialId).map(_ => initialId)(DEC)
-        }(DEC)
-
-    case None =>
-      logger.info("No ledger id given. Looking for existing ledger in database.")
-      ledgerDao
-        .lookupLedgerId()
-        .flatMap {
-          case Some(foundLedgerId) => ledgerFound(foundLedgerId)
-          case None =>
-            val randomLedgerId = LedgerIdGenerator.generateRandomId()
-            doInit(randomLedgerId).map(_ => randomLedgerId)(DEC)
-        }(DEC)
-  }
+  private def initialize(
+      initialLedgerId: Option[String],
+      initialLedgerEntries: ImmArray[LedgerEntryWithLedgerEndIncrement]): Future[String] =
+    initialLedgerId match {
+      case Some(initialId) =>
+        ledgerDao
+          .lookupLedgerId()
+          .flatMap {
+            case Some(foundLedgerId) if (foundLedgerId == initialId) =>
+              if (initialLedgerEntries.nonEmpty) {
+                logger.warn(
+                  s"Initial ledger entries provided, presumably from scenario, but I'm picking up from an existing database, and thus they will not be used")
+              }
+              ledgerFound(foundLedgerId)
+            case Some(foundLedgerId) =>
+              val errorMsg =
+                s"Ledger id mismatch. Ledger id given ('$initialId') is not equal to the existing one ('$foundLedgerId')!"
+              logger.error(errorMsg)
+              sys.error(errorMsg)
+            case None =>
+              if (initialLedgerEntries.nonEmpty) {
+                logger.info(
+                  s"Initializing ledger with ${initialLedgerEntries.length} ledger entries")
+              }
+              // Note that here we only store the ledger entry and we do not update anything else, such as the
+              // headRef. We also are not concerns with heartbeats / checkpoints. This is OK since this initialization
+              // step happens before we start up the sql ledger at all, so it's running in isolation.
+              @tailrec
```

Comment:
please factor this out into a separate method with a descriptive name

Reply:
why?

Reply:
Because it's good hygiene. `initialize` is getting too big and `go` (although a perfectly acceptable name for an internal tail-rec helper), doesn't tell anything about what this code is doing, hence one cannot have a quick overview of the pattern matching cases without understanding the recursive logic within.

Reply:
I think having it inside makes the class easier to read, since this method is _only used here_ and thus having it here removes the need for jumping around.

Regarding visualizing pattern matching, I've floated `go` to the top of the definition itself and renamed it.

Reply:
Ok, it looks better now.